Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs

```diff
@@ -0,0 +1,402 @@
+                m_ClientIDToGuid.Add(m_Portal.NetManager.LocalClientId, "host_guid");
+
+                AssignPlayerName(NetworkManager.Singleton.LocalClientId, m_Portal.PlayerName);
+            }
+        }
+
+        /// <summary>
+        /// Updates the player's session persistant data
+        /// </summary>
+        /// <param name="clientId">Id of the client to update</param>
+        /// <param name="playerName">The player's name to save</param>
+        /// <param name="avatarNetworkGuid">The NetworkGuid describing the player's class</param>
+        public void UpdatePlayerPersistantData(ulong clientId, string playerName, NetworkGuid avatarNetworkGuid)
+        {
+            if (m_ClientIDToGuid.TryGetValue(clientId, out var guid))
+            {
+                if (m_ClientData.TryGetValue(guid, out var sessionPlayerData))
+                {
+                    sessionPlayerData.PlayerName = playerName;
+                    sessionPlayerData.AvatarNetworkGuid = avatarNetworkGuid;
+                    m_ClientData[guid] = sessionPlayerData;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Updates the player's session data linked to their character
+        /// </summary>
+        /// <param name="clientId">Id of the client to update</param>
+        /// <param name="position">Current position of the player character</param>
+        /// <param name="rotation">Current rotation of the player character</param>
+        /// <param name="currentHp">Current hp of the player character</param>
+        public void UpdatePlayerCharacterData(ulong clientId, Vector3 position, Vector3 rotation, int currentHp)
+        {
+            if (m_ClientIDToGuid.TryGetValue(clientId, out var guid))
+            {
+                if (m_ClientData.TryGetValue(guid, out var sessionPlayerData))
+                {
+                    sessionPlayerData.PlayerPosition = position;
+                    sessionPlayerData.PlayerRotation = rotation;
+                    sessionPlayerData.CurrentHP = currentHp;
+                    m_ClientData[guid] = sessionPlayerData;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clears data from disconnected players, so that if they reconnect in the next game, they will be treated as new players
+        /// </summary>
+        public void OnGameEnded()
+        {
+            List<ulong> idsToClear = new List<ulong>();
+            foreach (var id in m_ClientIDToGuid.Keys)
+            {
+                if (!m_Portal.NetManager.ConnectedClientsIds.Contains(id))
+                {
+                    idsToClear.Add(id);
+                }
+            }
+
+            foreach (var id in idsToClear)
+            {
+                m_ClientData.Remove(m_ClientIDToGuid[id]);
+                m_ClientIDToGuid.Remove(id);
+            }
+        }
+
+        static void AssignPlayerName(ulong clientId, string playerName)
```

Comment:
I thought it was the session manager's responsibility to handle player data that it wants to keep and to reassign, but maybe it could simply carry the information instead. I suppose that I could remove these and make the PersistentPlayer assign them to itself, but I would need to make sure that it does so after the session manager has correctly updated its data.